R/domain.R

```diff
@@ -0,0 +1,81 @@
+#' @include utils-pipe.R
+NULL
+
+domain <- function(client, domain_name) {
+  protocol <- client$.__protocol__
+  members_names <- names(c(
+    protocol$list_commands(domain_name),
+    protocol$list_events(domain_name),
+    protocol$list_types(domain_name)
+  ))
+
+  CustomDomain <- R6::R6Class(
+    domain_name,
+    inherit = Domain,
+    public = list(
+      initialize = function(client, domain) {
+        super$initialize(client, domain)
+        protocol <- client$.__protocol__
+        commands <- protocol$list_commands(domain)
+        events <- protocol$list_events(domain)
+        types <- protocol$list_types(domain)
+        purrr::iwalk(commands, function(command, name) {
+          self[[name]] <- private$.build_command(command, name)
+        })
+        purrr::iwalk(events, function(event, name) {
+          self[[name]] <- private$.build_event_listener(event)
+        })
+      }
+    )
+  )
+
+  purrr::walk(members_names, ~ CustomDomain$set("public", .x, NULL))
+  CustomDomain$new(client, domain_name)
+}
+
+Domain <- R6::R6Class(
+  "Domain",
+  public = list(
+    initialize = function(client, domain) {
+      self$.__client__ <- client
+      private$.domain_name <- domain
+    },
+    .__client__ = NULL
+  ),
+  private = list(
+    .domain_name = NULL,
+    .build_command = function(method_to_be_sent, name) {
+      fun <- function() {
+        rlang::fn_fmls_names() %>% # pick the fun arguments
+          utils::head(-1) %>% # remove the callback argument
+          rlang::env_get_list(nms = ., inherit = TRUE) %>% # retrieve values
+          purrr::discard(~ purrr::is_null(.x)) %>% # remove arguments identical to NULL
```

Comment:
For easier debugging, I would stop the pipe here to assign the result to a list and pass it into `send` `params`
```r
params  <-  rlang::fn_fmls_names() %>% # pick the fun arguments
          utils::head(-1) %>% # remove the callback argument
          rlang::env_get_list(nms = ., inherit = TRUE) %>% # retrieve values
          purrr::discard(~ purrr::is_null(.x))
 self$.__client__$send(get("method_to_be_sent", envir = parent.env(environment())), 
      params = params, onresponse = callback)
```
It should not change much but would allow to go inside send when debugging interactively. 
I don't think naming `params` object would conflict

Reply:
I think `params` is risky: if we use a snake case name, I think this will be secure (because CDP use camel case)

Reply:
You are right.  `params_to_be_sent` ? 😉

Reply:
done in https://github.com/RLesur/crrri/pull/40/commits/ed5cbe00e0590edc4430255591efc1f1baecce83